relay/pkg/chain/ethereum/ethereum.go

```diff
@@ -11,26 +20,220 @@ import (
 
 var logger = log.Logger("relay-chain-ethereum")
 
+// RelayContractName defines the name of the Relay contract.
+const RelayContractName = "Relay"
+
+var (
+	// DefaultMiningCheckInterval is the default interval in which transaction
+	// mining status is checked. If the transaction is not mined within this
+	// time, the gas price is increased and transaction is resubmitted.
+	// This value can be overwritten in the configuration file.
+	DefaultMiningCheckInterval = 60 * time.Second
+
+	// DefaultMaxGasPrice specifies the default maximum gas price the client is
+	// willing to pay for the transaction to be mined. The offered transaction
+	// gas price can not be higher than the max gas price value. If the maximum
+	// allowed gas price is reached, no further resubmission attempts are
+	// performed. This value can be overwritten in the configuration file.
+	DefaultMaxGasPrice = big.NewInt(1000000000000) // 1000 Gwei
+)
+
 // ethereumChain is an implementation of the host chain interface for Ethereum.
 type ethereumChain struct {
-	// TODO: implementation
+	config        *ethereum.Config
+	accountKey    *keystore.Key
+	client        ethutil.EthereumClient
+	relayContract *contract.Relay
+	blockCounter  *ethlike.BlockCounter
+	miningWaiter  *ethlike.MiningWaiter
+	nonceManager  *ethlike.NonceManager
+
+	// transactionMutex allows interested parties to forcibly serialize
+	// transaction submission.
+	//
+	// When transactions are submitted, they require a valid nonce. The nonce is
+	// equal to the count of transactions the account has submitted so far, and
+	// for a transaction to be accepted it should be monotonically greater than
+	// any previous submitted transaction. To do this, transaction submission
+	// asks the Ethereum client it is connected to for the next pending nonce,
+	// and uses that value for the transaction. Unfortunately, if multiple
+	// transactions are submitted in short order, they may all get the same
+	// nonce. Serializing submission ensures that each nonce is requested after
+	// a previous transaction has been submitted.
+	transactionMutex *sync.Mutex
 }
 
 // Connect performs initialization for communication with Ethereum blockchain
 // based on provided config.
 func Connect(
-	ctx context.Context,
+	_ context.Context,
```

Comment:
Let's remove this param if we don't use it

Reply:
Good catch! Please see https://github.com/keep-network/tbtc/commit/e9d58c2a36743b880bba8d1182e198361e4034ad